appserver/tests/embedded/scatteredarchive/src/test/java/org/glassfish/tests/embedded/scatteredarchive/ScatteredArchiveTest.java

```diff
@@ -108,42 +135,91 @@ public void test() throws Exception {
+    }
+
+    @Test
+    public void testContextInitialized() throws Exception {
+        Class<ContextInitializedTestServlet> servletClass = ContextInitializedTestServlet.class;
+        String ARCHIVE_NAME = servletClass.getSimpleName() + "Archive";
+
+        ScatteredArchive sa = createDefaultArchive(ARCHIVE_NAME);
+        String testClassesSubPath = servletClass.getPackageName().replace('.', File.separatorChar);
+        String additionalClassPath = "target/" + servletClass.getSimpleName() + "-classes/";
+
+        // copy test-specific classes and add them to the archive
+        FileUtils.copyTree(new File(PROJECT_DIR, "target/test-classes/" + testClassesSubPath),
+                new File(PROJECT_DIR, additionalClassPath + testClassesSubPath));
+        sa.addClassPath(new File(PROJECT_DIR, additionalClassPath));
+
+        // add some JAR files to the archive
+        sa.addClassPath(new File(PROJECT_DIR, "target/test-dependencies/hamcrest.jar"));
+        sa.addClassPath(new File(PROJECT_DIR, "target/test-dependencies/junit-jupiter-engine.jar"));
 
+        URI warURI = sa.toURI();
+        printContents(warURI);
+
+        // Deploy archive
+        Deployer deployer = glassfish.getDeployer();
+        String appname = deployer.deploy(warURI);
+        System.out.println("Deployed [" + appname + "]");
+        Assertions.assertEquals(appname, ARCHIVE_NAME);
+
+        get("http://localhost:8080/satest/" + ContextInitializedTestServlet.class.getSimpleName(),
+                LABEL_CONTEXT_INITIALIZED_COUNTER, "1");
+    }
+
+    private ScatteredArchive createDefaultArchive(String ARCHIVE_NAME) throws IOException {
+        // Test Scattered Web Archive
+        ScatteredArchive sa = new ScatteredArchive(ARCHIVE_NAME,
+                ScatteredArchive.Type.WAR, new File(PROJECT_DIR, "src/main/webapp"));
+        sa.addClassPath(new File(PROJECT_DIR, "target/classes"));
+        sa.addClassPath(new File(PROJECT_DIR, "src/main/resources"));
+        return sa;
+    }
+
+    private void get(String urlStr, String containingString) throws Exception {
+        List<String> inLines = getLinesFromUrl(new URL(urlStr));
+        MatcherAssert.assertThat("Output from servlet", inLines, hasItem(containsString(containingString)));
+        logger.log(INFO, "***** SUCCESS **** Found [" + containingString + "] in the response.*****");
+    }
+
+    private void get(String urlStr, String key, String value) throws Exception {
+        List<String> inLines = getLinesFromUrl(new URL(urlStr));
+        String result = key + ":" + value;
+        MatcherAssert.assertThat("Output from servlet", inLines, hasItem(is(result)));
+        logger.log(INFO, "***** SUCCESS **** Found [" + result + "] in the response.*****");
     }
 
-    private void get(String urlStr, String result) throws Exception {
-        URL url = new URL(urlStr);
+    private List<String> getLinesFromUrl(URL url) throws Exception {
         URLConnection yc = url.openConnection();
```

Comment:
Wrap with try and disconnect in finally?

Reply:
There's no disconnect on URLConnection. The connection isn't created with `URL.openConnection()`. It's created later, with `yc.getInputStream()`, see the Javadoc: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/net/URL.html#openConnection()

Reply:
You're right. Sorry :)

Reply:
No problem. Thanks for being so thorough :)